{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "import itertools\n",
    "import os\n",
    "warnings.filterwarnings('ignore')\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gseapy.parser import Biomart\n",
    "\n",
    "bm = Biomart(verbose=False, host=\"europe.ensembl.org\")\n",
    "## view validated marts\n",
    "marts = bm.get_marts()\n",
    "## view validated dataset\n",
    "datasets = bm.get_datasets(mart='ENSEMBL_MART_ENSEMBL')\n",
    "## view validated attributes\n",
    "attrs = bm.get_attributes(dataset='hsapiens_gene_ensembl')\n",
    "## view validated filters\n",
    "filters = bm.get_filters(dataset='hsapiens_gene_ensembl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSEMBL_MART_ENSEMBL</td>\n",
       "      <td>Ensembl Genes 103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSEMBL_MART_MOUSE</td>\n",
       "      <td>Mouse strains 103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSEMBL_MART_SEQUENCE</td>\n",
       "      <td>Sequence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSEMBL_MART_ONTOLOGY</td>\n",
       "      <td>Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSEMBL_MART_GENOMIC</td>\n",
       "      <td>Genomic features 103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSEMBL_MART_SNP</td>\n",
       "      <td>Ensembl Variation 103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSEMBL_MART_FUNCGEN</td>\n",
       "      <td>Ensembl Regulation 103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Name             Description\n",
       "0   ENSEMBL_MART_ENSEMBL       Ensembl Genes 103\n",
       "1     ENSEMBL_MART_MOUSE       Mouse strains 103\n",
       "2  ENSEMBL_MART_SEQUENCE                Sequence\n",
       "3  ENSEMBL_MART_ONTOLOGY                Ontology\n",
       "4   ENSEMBL_MART_GENOMIC    Genomic features 103\n",
       "5       ENSEMBL_MART_SNP   Ensembl Variation 103\n",
       "6   ENSEMBL_MART_FUNCGEN  Ensembl Regulation 103"
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "marts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add the names of the pathway for the DKO drugs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {},
   "outputs": [],
   "source": [
    "def explode(df, lst_cols, fill_value='', preserve_index=False):\n",
    "    # make sure `lst_cols` is list-alike\n",
    "    if (lst_cols is not None\n",
    "        and len(lst_cols) > 0\n",
    "        and not isinstance(lst_cols, (list, tuple, np.ndarray, pd.Series))):\n",
    "        lst_cols = [lst_cols]\n",
    "    # all columns except `lst_cols`\n",
    "    idx_cols = df.columns.difference(lst_cols)\n",
    "    # calculate lengths of lists\n",
    "    lens = df[lst_cols[0]].str.len()\n",
    "    # preserve original index values    \n",
    "    idx = np.repeat(df.index.values, lens)\n",
    "    # create \"exploded\" DF\n",
    "    res = (pd.DataFrame({\n",
    "                col:np.repeat(df[col].values, lens)\n",
    "                for col in idx_cols},\n",
    "                index=idx)\n",
    "             .assign(**{col:np.concatenate(df.loc[lens>0, col].values)\n",
    "                            for col in lst_cols}))\n",
    "    # append those rows that have empty lists\n",
    "    if (lens == 0).any():\n",
    "        # at least one list in cells is empty\n",
    "        res = (res.append(df.loc[lens==0, idx_cols], sort=False)\n",
    "                  .fillna(fill_value))\n",
    "    # revert the original index order\n",
    "    res = res.sort_index()\n",
    "    # reset index if requested\n",
    "    if not preserve_index:        \n",
    "        res = res.reset_index(drop=True)\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Genes</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Purine metabolism</td>\n",
       "      <td>[RRM1, RRM2, NME4, PAICS, PFAS, PNP, ATIC, RRM...</td>\n",
       "      <td>1.790315e-16</td>\n",
       "      <td>13/129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Citrate cycle (TCA cycle)</td>\n",
       "      <td>[FH, SUCLG2, SDHC, SUCLG1, SDHD, SDHA, SDHB, DLD]</td>\n",
       "      <td>3.866059e-14</td>\n",
       "      <td>8/30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>[DTYMK, RRM1, PNP, RRM2, RRM2B, DPYD, CMPK1, N...</td>\n",
       "      <td>1.020544e-11</td>\n",
       "      <td>8/57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>[RPIA, GPI, PGM2, PGD, DERA]</td>\n",
       "      <td>3.832756e-08</td>\n",
       "      <td>5/30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Drug metabolism</td>\n",
       "      <td>[RRM1, RRM2, RRM2B, DPYD, CMPK1, NME4, HPRT1]</td>\n",
       "      <td>5.304212e-08</td>\n",
       "      <td>7/108</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Term  \\\n",
       "0          Purine metabolism   \n",
       "1  Citrate cycle (TCA cycle)   \n",
       "2      Pyrimidine metabolism   \n",
       "3  Pentose phosphate pathway   \n",
       "4            Drug metabolism   \n",
       "\n",
       "                                               Genes       P-value Overlap  \n",
       "0  [RRM1, RRM2, NME4, PAICS, PFAS, PNP, ATIC, RRM...  1.790315e-16  13/129  \n",
       "1  [FH, SUCLG2, SDHC, SUCLG1, SDHD, SDHA, SDHB, DLD]  3.866059e-14    8/30  \n",
       "2  [DTYMK, RRM1, PNP, RRM2, RRM2B, DPYD, CMPK1, N...  1.020544e-11    8/57  \n",
       "3                       [RPIA, GPI, PGM2, PGD, DERA]  3.832756e-08    5/30  \n",
       "4      [RRM1, RRM2, RRM2B, DPYD, CMPK1, NME4, HPRT1]  5.304212e-08   7/108  "
      ]
     },
     "execution_count": 320,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read Enrichr DKO enrichment\n",
    "dko_enrichr = pd.read_csv('KO_data/DKO_Lung_Studies_Enrichment.csv',index_col=0)\n",
    "dko_enrichr.Genes = dko_enrichr.Genes.apply(lambda x:x.split(';'))\n",
    "dko_enrichr_ = explode(dko_enrichr,['Genes'], fill_value='')\n",
    "dko_enrichr_ = dko_enrichr_.loc[:,['Genes','Term']].drop_duplicates().reset_index(drop=True)\n",
    "dko_enrichr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway</th>\n",
       "      <th>grRules</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14482</th>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>1351.1 and 1347.1 and 1329.1 and 1327.1 and 34...</td>\n",
       "      <td>1340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14483</th>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>(7384.1 and 7388.1 and 4519.1 and 10975.1 and ...</td>\n",
       "      <td>1340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14484</th>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>1351.1 and 1347.1 and 1329.1 and 1327.1 and 34...</td>\n",
       "      <td>84701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14485</th>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>1351.1 and 1347.1 and 1329.1 and 1327.1 and 34...</td>\n",
       "      <td>1337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14486</th>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "      <td>(7384.1 and 7388.1 and 4519.1 and 10975.1 and ...</td>\n",
       "      <td>1337</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Pathway  \\\n",
       "14482  Oxidative phosphorylation   \n",
       "14483  Oxidative phosphorylation   \n",
       "14484  Oxidative phosphorylation   \n",
       "14485  Oxidative phosphorylation   \n",
       "14486  Oxidative phosphorylation   \n",
       "\n",
       "                                                 grRules   gene  \n",
       "14482  1351.1 and 1347.1 and 1329.1 and 1327.1 and 34...   1340  \n",
       "14483  (7384.1 and 7388.1 and 4519.1 and 10975.1 and ...   1340  \n",
       "14484  1351.1 and 1347.1 and 1329.1 and 1327.1 and 34...  84701  \n",
       "14485  1351.1 and 1347.1 and 1329.1 and 1327.1 and 34...   1337  \n",
       "14486  (7384.1 and 7388.1 and 4519.1 and 10975.1 and ...   1337  "
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read RECON3 pwthways & rules to do enrichment for DKO drugs\n",
    "recon_rules = pd.read_csv('RECON3_subSystems_grRules_genes.csv')\n",
    "recon_rules.gene = recon_rules.gene.apply(lambda x:str(str(x).split('.')[0]))\n",
    "recon_rules.tail(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway</th>\n",
       "      <th>grRules</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Beta-Alanine metabolism</td>\n",
       "      <td>8639.1 or 26.1 or 314.2 or 314.1</td>\n",
       "      <td>8639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "      <td>4129.1 or 8639.1 or 26.1 or 314.2 or 314.1 or ...</td>\n",
       "      <td>8639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Glycine, serine, alanine, and threonine metabo...</td>\n",
       "      <td>8639.1</td>\n",
       "      <td>8639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "      <td>8639.1 or 26.1 or 314.2 or 314.1</td>\n",
       "      <td>8639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>26.1 or 314.1 or 8639.1 or 314.2</td>\n",
       "      <td>8639</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Pathway  \\\n",
       "0                            Beta-Alanine metabolism   \n",
       "1                                Tyrosine metabolism   \n",
       "2  Glycine, serine, alanine, and threonine metabo...   \n",
       "3                                Tyrosine metabolism   \n",
       "4                               Histidine metabolism   \n",
       "\n",
       "                                             grRules  gene  \n",
       "0                   8639.1 or 26.1 or 314.2 or 314.1  8639  \n",
       "1  4129.1 or 8639.1 or 26.1 or 314.2 or 314.1 or ...  8639  \n",
       "2                                             8639.1  8639  \n",
       "3                   8639.1 or 26.1 or 314.2 or 314.1  8639  \n",
       "4                   26.1 or 314.1 or 8639.1 or 314.2  8639  "
      ]
     },
     "execution_count": 322,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recon_rules.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = recon_rules.gene.unique().tolist()\n",
    "results = bm.query(dataset='hsapiens_gene_ensembl', attributes=['entrezgene_id','hgnc_symbol'],#, 'go_id'],\n",
    "                      filters={'entrezgene_id': gene_list})\n",
    "results =  results.dropna().drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# recon_rules = recon_rules.dropna()\n",
    "recon_rules.grRules = recon_rules.grRules.apply(lambda x:x.replace(' or ',';').replace(' and ',';'))\n",
    "recon_rules.grRules = recon_rules.grRules.apply(lambda x:x.replace('(','').replace(')',''))\n",
    "recon_rules.grRules = recon_rules.grRules.apply(lambda x:x.split(';'))\n",
    "recon_rules = explode(recon_rules,['grRules'], fill_value='')\n",
    "recon_rules.grRules = recon_rules.grRules.apply(lambda x:x.split('.')[0])\n",
    "recon_rules = recon_rules.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14487, 3)"
      ]
     },
     "execution_count": 324,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recon_rules.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1872, 2)"
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1884"
      ]
     },
     "execution_count": 326,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [],
   "source": [
    "recon_rules['Genes'] = 0\n",
    "for i in range(recon_rules.shape[0]):\n",
    "    entrez = recon_rules.iloc[i,2]\n",
    "    if int(entrez) in results.entrezgene_id.values:\n",
    "        gene = set(results.loc[results.entrezgene_id == int(entrez),'hgnc_symbol'].values.tolist())\n",
    "    else:\n",
    "        gene = np.NAN\n",
    "    recon_rules.iloc[i,3] =  gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway</th>\n",
       "      <th>grRules</th>\n",
       "      <th>gene</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Beta-Alanine metabolism</td>\n",
       "      <td>8639.1 or 26.1 or 314.2 or 314.1</td>\n",
       "      <td>8639</td>\n",
       "      <td>AOC3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "      <td>4129.1 or 8639.1 or 26.1 or 314.2 or 314.1 or ...</td>\n",
       "      <td>8639</td>\n",
       "      <td>{AOC3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Glycine, serine, alanine, and threonine metabo...</td>\n",
       "      <td>8639.1</td>\n",
       "      <td>8639</td>\n",
       "      <td>{AOC3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "      <td>8639.1 or 26.1 or 314.2 or 314.1</td>\n",
       "      <td>8639</td>\n",
       "      <td>{AOC3}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>26.1 or 314.1 or 8639.1 or 314.2</td>\n",
       "      <td>8639</td>\n",
       "      <td>{AOC3}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Pathway  \\\n",
       "0                            Beta-Alanine metabolism   \n",
       "1                                Tyrosine metabolism   \n",
       "2  Glycine, serine, alanine, and threonine metabo...   \n",
       "3                                Tyrosine metabolism   \n",
       "4                               Histidine metabolism   \n",
       "\n",
       "                                             grRules  gene   Genes  \n",
       "0                   8639.1 or 26.1 or 314.2 or 314.1  8639    AOC3  \n",
       "1  4129.1 or 8639.1 or 26.1 or 314.2 or 314.1 or ...  8639  {AOC3}  \n",
       "2                                             8639.1  8639  {AOC3}  \n",
       "3                   8639.1 or 26.1 or 314.2 or 314.1  8639  {AOC3}  \n",
       "4                   26.1 or 314.1 or 8639.1 or 314.2  8639  {AOC3}  "
      ]
     },
     "execution_count": 328,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recon_rules.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "metadata": {},
   "outputs": [],
   "source": [
    "recon_rules = recon_rules.loc[:,['Pathway','Genes']].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [],
   "source": [
    "recon_rules.Genes = recon_rules.Genes.apply(lambda x:''.join(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Beta-Alanine metabolism</td>\n",
       "      <td>AOC3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "      <td>AOC3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Glycine, serine, alanine, and threonine metabo...</td>\n",
       "      <td>AOC3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "      <td>AOC3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>AOC3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Pathway Genes\n",
       "0                            Beta-Alanine metabolism  AOC3\n",
       "1                                Tyrosine metabolism  AOC3\n",
       "2  Glycine, serine, alanine, and threonine metabo...  AOC3\n",
       "3                                Tyrosine metabolism  AOC3\n",
       "4                               Histidine metabolism  AOC3"
      ]
     },
     "execution_count": 331,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recon_rules.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14417, 2)"
      ]
     },
     "execution_count": 332,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recon_rules.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug</th>\n",
       "      <th>Number_Genes</th>\n",
       "      <th>Phenotype_Essentiality</th>\n",
       "      <th>Phenotype_Safety</th>\n",
       "      <th>Gene_Pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CMPK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SLC29A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TYMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SLC29A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SLC29A2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Drug  Number_Genes  Phenotype_Essentiality  Phenotype_Safety  \\\n",
       "0  Gemcitabine             3                    1.66               1.0   \n",
       "1  Gemcitabine             3                    1.66               1.0   \n",
       "2  Gemcitabine             3                    1.66               1.0   \n",
       "3  Gemcitabine             3                    1.66               1.0   \n",
       "4  Gemcitabine             3                    1.66               1.0   \n",
       "\n",
       "  Gene_Pair  \n",
       "0     CMPK1  \n",
       "1   SLC29A2  \n",
       "2      TYMS  \n",
       "3   SLC29A2  \n",
       "4   SLC29A2  "
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add the names of the pathway for the DKO drugs\n",
    "dko_one_drug = pd.read_csv('KO_data/DKO_Drugs_OneDrug.csv',index_col=0)\n",
    "dko_one_drug.Gene_Pair = dko_one_drug.Gene_Pair.apply(lambda x:str(x).replace('{','').replace('}','').replace(', ',';').replace(\"'\",''))\n",
    "dko_one_drug.Gene_Pair = dko_one_drug.Gene_Pair.apply(lambda x:x.split(';'))\n",
    "dko_one_drug_ =  explode(dko_one_drug,['Gene_Pair'], fill_value='')\n",
    "dko_one_drug_.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug</th>\n",
       "      <th>Number_Genes</th>\n",
       "      <th>Phenotype_Essentiality</th>\n",
       "      <th>Phenotype_Safety</th>\n",
       "      <th>Gene_Pair</th>\n",
       "      <th>Pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CMPK1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SLC29A2</td>\n",
       "      <td>Transport, extracellular;Transport, extracellu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TYMS</td>\n",
       "      <td>Nucleotide interconversion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SLC29A2</td>\n",
       "      <td>Transport, extracellular;Transport, extracellu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SLC29A2</td>\n",
       "      <td>Transport, extracellular;Transport, extracellu...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Drug  Number_Genes  Phenotype_Essentiality  Phenotype_Safety  \\\n",
       "0  Gemcitabine             3                    1.66               1.0   \n",
       "1  Gemcitabine             3                    1.66               1.0   \n",
       "2  Gemcitabine             3                    1.66               1.0   \n",
       "3  Gemcitabine             3                    1.66               1.0   \n",
       "4  Gemcitabine             3                    1.66               1.0   \n",
       "\n",
       "  Gene_Pair                                           Pathways  \n",
       "0     CMPK1  Nucleotide interconversion;Nucleotide intercon...  \n",
       "1   SLC29A2  Transport, extracellular;Transport, extracellu...  \n",
       "2      TYMS                         Nucleotide interconversion  \n",
       "3   SLC29A2  Transport, extracellular;Transport, extracellu...  \n",
       "4   SLC29A2  Transport, extracellular;Transport, extracellu...  "
      ]
     },
     "execution_count": 334,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes =  dko_one_drug_.Gene_Pair.unique()\n",
    "dko_one_drug_['Pathways'] = 0\n",
    "for gene in genes:\n",
    "    gene_pathways = recon_rules.loc[recon_rules.Genes ==gene,'Pathway'].values.tolist()\n",
    "    dko_one_drug_.loc[dko_one_drug_.Gene_Pair ==gene,'Pathways'] = ';'.join(gene_pathways)\n",
    "dko_one_drug_.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drug</th>\n",
       "      <th>Number_Genes</th>\n",
       "      <th>Phenotype_Essentiality</th>\n",
       "      <th>Phenotype_Safety</th>\n",
       "      <th>Gene_Pair</th>\n",
       "      <th>Pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CMPK1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SLC29A2</td>\n",
       "      <td>Transport, extracellular;Transport, extracellu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TYMS</td>\n",
       "      <td>Nucleotide interconversion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.66</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SLC29A1</td>\n",
       "      <td>Transport, extracellular;Transport, mitochondr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Trifluridine</td>\n",
       "      <td>3</td>\n",
       "      <td>1.33</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TYMS</td>\n",
       "      <td>Nucleotide interconversion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Drug  Number_Genes  Phenotype_Essentiality  Phenotype_Safety  \\\n",
       "0   Gemcitabine             3                    1.66               1.0   \n",
       "1   Gemcitabine             3                    1.66               1.0   \n",
       "2   Gemcitabine             3                    1.66               1.0   \n",
       "3   Gemcitabine             3                    1.66               1.0   \n",
       "4  Trifluridine             3                    1.33               1.0   \n",
       "\n",
       "  Gene_Pair                                           Pathways  \n",
       "0     CMPK1  Nucleotide interconversion;Nucleotide intercon...  \n",
       "1   SLC29A2  Transport, extracellular;Transport, extracellu...  \n",
       "2      TYMS                         Nucleotide interconversion  \n",
       "3   SLC29A1  Transport, extracellular;Transport, mitochondr...  \n",
       "4      TYMS                         Nucleotide interconversion  "
      ]
     },
     "execution_count": 335,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dko_one_drug_path = dko_one_drug_.loc[dko_one_drug_.Pathways.apply(lambda x:len(x))>0,:]\n",
    "dko_one_drug_path = dko_one_drug_path.drop_duplicates().reset_index(drop=True)\n",
    "dko_one_drug_path.to_csv('KO_data/DKO_Drugs_OneDrug_Pathways.csv')\n",
    "dko_one_drug_path.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [],
   "source": [
    "dko_one_drug_path.Pathways = dko_one_drug_path.Pathways.apply(lambda x:x.split(';'))\n",
    "dko_one_drug_path_ = explode(dko_one_drug_path,['Pathways'],'')\n",
    "dko_one_drug_path_ = dko_one_drug_path_.drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "dko_one_drug_path_.to_csv('KO_data/DKO_Drugs_OneDrug_Pathways_tripartite.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drugs1</th>\n",
       "      <th>Drugs2</th>\n",
       "      <th>Number_Genes</th>\n",
       "      <th>Phenotype_Essentiality</th>\n",
       "      <th>Phenotype_Safety</th>\n",
       "      <th>Gene_Pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GUK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RRM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GUK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RRM2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Acyclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GUK1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Drugs1        Drugs2  Number_Genes  Phenotype_Essentiality  \\\n",
       "0  Imexon  Valaciclovir             2                    10.0   \n",
       "1  Imexon  Valaciclovir             2                    10.0   \n",
       "2  Imexon  Valaciclovir             2                    10.0   \n",
       "3  Imexon  Valaciclovir             2                    10.0   \n",
       "4  Imexon     Acyclovir             2                    10.0   \n",
       "\n",
       "   Phenotype_Safety Gene_Pair  \n",
       "0               0.0      GUK1  \n",
       "1               0.0      RRM1  \n",
       "2               0.0      GUK1  \n",
       "3               0.0      RRM2  \n",
       "4               0.0      GUK1  "
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add the names of the pathway for the DKO drugs\n",
    "dko_drug = pd.read_csv('KO_data/DKO_Drug_Repurposing_Reduced.csv',index_col=0)\n",
    "dko_drug.Gene_Pair = dko_drug.Gene_Pair.apply(lambda x:str(x).replace('{','').replace('}','').replace(', ',';').replace(\"'\",''))\n",
    "dko_drug.Gene_Pair = dko_drug.Gene_Pair.apply(lambda x:x.split(';'))\n",
    "dko_drug_ =  explode(dko_drug,['Gene_Pair'], fill_value='')\n",
    "dko_drug_.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drugs1</th>\n",
       "      <th>Drugs2</th>\n",
       "      <th>Number_Genes</th>\n",
       "      <th>Phenotype_Essentiality</th>\n",
       "      <th>Phenotype_Safety</th>\n",
       "      <th>Gene_Pair</th>\n",
       "      <th>Pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RRM1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RRM2</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Acyclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Drugs1        Drugs2  Number_Genes  Phenotype_Essentiality  \\\n",
       "0  Imexon  Valaciclovir             2                    10.0   \n",
       "1  Imexon  Valaciclovir             2                    10.0   \n",
       "2  Imexon  Valaciclovir             2                    10.0   \n",
       "3  Imexon  Valaciclovir             2                    10.0   \n",
       "4  Imexon     Acyclovir             2                    10.0   \n",
       "\n",
       "   Phenotype_Safety Gene_Pair  \\\n",
       "0               0.0      GUK1   \n",
       "1               0.0      RRM1   \n",
       "2               0.0      GUK1   \n",
       "3               0.0      RRM2   \n",
       "4               0.0      GUK1   \n",
       "\n",
       "                                            Pathways  \n",
       "0  Nucleotide interconversion;Nucleotide intercon...  \n",
       "1  Nucleotide interconversion;Nucleotide intercon...  \n",
       "2  Nucleotide interconversion;Nucleotide intercon...  \n",
       "3  Nucleotide interconversion;Nucleotide intercon...  \n",
       "4  Nucleotide interconversion;Nucleotide intercon...  "
      ]
     },
     "execution_count": 338,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes =  dko_drug_.Gene_Pair.unique()\n",
    "dko_drug_['Pathways'] = 0\n",
    "for gene in genes:\n",
    "    gene_pathways = recon_rules.loc[recon_rules.Genes ==gene,'Pathway'].values.tolist()\n",
    "    dko_drug_.loc[dko_drug_.Gene_Pair ==gene,'Pathways'] = ';'.join(gene_pathways)\n",
    "dko_drug_.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drugs1</th>\n",
       "      <th>Drugs2</th>\n",
       "      <th>Number_Genes</th>\n",
       "      <th>Phenotype_Essentiality</th>\n",
       "      <th>Phenotype_Safety</th>\n",
       "      <th>Gene_Pair</th>\n",
       "      <th>Pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RRM1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RRM2</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Acyclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Imexon</td>\n",
       "      <td>Acyclovir</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RRM1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Drugs1        Drugs2  Number_Genes  Phenotype_Essentiality  \\\n",
       "0  Imexon  Valaciclovir             2                    10.0   \n",
       "1  Imexon  Valaciclovir             2                    10.0   \n",
       "2  Imexon  Valaciclovir             2                    10.0   \n",
       "3  Imexon     Acyclovir             2                    10.0   \n",
       "4  Imexon     Acyclovir             2                    10.0   \n",
       "\n",
       "   Phenotype_Safety Gene_Pair  \\\n",
       "0               0.0      GUK1   \n",
       "1               0.0      RRM1   \n",
       "2               0.0      RRM2   \n",
       "3               0.0      GUK1   \n",
       "4               0.0      RRM1   \n",
       "\n",
       "                                            Pathways  \n",
       "0  Nucleotide interconversion;Nucleotide intercon...  \n",
       "1  Nucleotide interconversion;Nucleotide intercon...  \n",
       "2  Nucleotide interconversion;Nucleotide intercon...  \n",
       "3  Nucleotide interconversion;Nucleotide intercon...  \n",
       "4  Nucleotide interconversion;Nucleotide intercon...  "
      ]
     },
     "execution_count": 339,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dko_drug_path = dko_drug_.loc[dko_drug_.Pathways.apply(lambda x:len(x))>0,:]\n",
    "dko_drug_path = dko_drug_path.drop_duplicates().reset_index(drop=True)\n",
    "dko_drug_path.to_csv('KO_data/DKO_Drug_Repurposing_Reduced_Pathways.csv')\n",
    "dko_drug_path.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {},
   "outputs": [],
   "source": [
    "dko_drug_path.Pathways = dko_drug_path.Pathways.apply(lambda x:x.split(';'))\n",
    "dko_drug_path_ = explode(dko_drug_path,['Pathways'],'')\n",
    "dko_drug_path_ = dko_drug_path_.drop_duplicates().reset_index(drop=True)\n",
    "dko_drug_path_.to_csv('KO_data/DKO_Drug_Repurposing_Reduced_Pathways_tripartite.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [],
   "source": [
    "sko_drugs = pd.read_csv('KO_data/SKO_db_Drugs_Unique.csv')\n",
    "sko_drugs.Gene_Symbol = sko_drugs.Gene_Symbol.apply(lambda x:str(x[2:len(x)-2]))\n",
    "genes =  sko_drugs.Gene_Symbol.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "\n",
    "enr = gp.enrichr(gene_list=genes,#dko_all_genes,\n",
    "                 gene_sets=['KEGG_2019_Human'],\n",
    "                 cutoff=1 # test dataset, use lower value from range(0,1)\n",
    "                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrich = enr.results\n",
    "enrich = enrich.loc[:,['Term','Genes','P-value','Overlap']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DHFR', 'SLC7A11', 'SLC7A5', 'ISYNA1', 'CMPK1', 'GUK1', 'SLC7A5']"
      ]
     },
     "execution_count": 344,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x for x in enrich.Genes if x in genes ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GUK1', 'CMPK1', 'CMPK1', 'SLC7A11', 'SLC7A5', 'SLC7A5', 'ISYNA1']"
      ]
     },
     "execution_count": 345,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x for x in dko_enrichr_.Genes if x in genes ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drugs</th>\n",
       "      <th>Drugs_Categories</th>\n",
       "      <th>Drugs_Group</th>\n",
       "      <th>Number_Genes</th>\n",
       "      <th>Phenotype_Essentiality</th>\n",
       "      <th>Phenotype_Safety</th>\n",
       "      <th>side_effect_name</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>Acyclovir and prodrug|Amino Acids|Amino Acids,...</td>\n",
       "      <td>['approved|investigational']</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acyclovir</td>\n",
       "      <td>Acyclovir and prodrug|Anti-Infective Agents|An...</td>\n",
       "      <td>['approved']</td>\n",
       "      <td>1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>Nucleotide interconversion;Nucleotide intercon...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Drugs                                   Drugs_Categories  \\\n",
       "0  Valaciclovir  Acyclovir and prodrug|Amino Acids|Amino Acids,...   \n",
       "1     Acyclovir  Acyclovir and prodrug|Anti-Infective Agents|An...   \n",
       "\n",
       "                    Drugs_Group  Number_Genes  Phenotype_Essentiality  \\\n",
       "0  ['approved|investigational']             1                    16.0   \n",
       "1                  ['approved']             1                    16.0   \n",
       "\n",
       "   Phenotype_Safety side_effect_name Gene_Symbol  \\\n",
       "0               3.0            [nan]        GUK1   \n",
       "1               3.0            [nan]        GUK1   \n",
       "\n",
       "                                            Pathways  \n",
       "0  Nucleotide interconversion;Nucleotide intercon...  \n",
       "1  Nucleotide interconversion;Nucleotide intercon...  "
      ]
     },
     "execution_count": 346,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Map SKO drugs to drug-gene-pathway\n",
    "sko_drugs = pd.read_csv('KO_data/SKO_db_Drugs_Unique.csv')\n",
    "sko_drugs.Gene_Symbol = sko_drugs.Gene_Symbol.apply(lambda x:str(x[2:len(x)-2]).split(\"\"\"', '\"\"\"))#list(str(x[2:len(x)-2]))#.replace(\"\"\"', '\"\"\",';')\n",
    "sko_drugs = explode(sko_drugs,['Gene_Symbol'],'')\n",
    "\n",
    "genes =  sko_drugs.Gene_Symbol.unique().tolist()\n",
    "sko_drugs['Pathways'] = 0\n",
    "for gene in genes:\n",
    "    gene_pathways = recon_rules.loc[recon_rules.Genes ==gene,'Pathway'].values.tolist()\n",
    "    sko_drugs.loc[sko_drugs.Gene_Symbol ==gene,'Pathways'] = ';'.join(gene_pathways)\n",
    "sko_drugs.to_csv('KO_data/SKO_db_Drugs_Unique_Pathways.csv')\n",
    "sko_drugs.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Drugs</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Phenotype_Essentiality</th>\n",
       "      <th>Phenotype_Safety</th>\n",
       "      <th>Pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Valaciclovir</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Nucleotide interconversion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acyclovir</td>\n",
       "      <td>GUK1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>Nucleotide interconversion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Drugs Gene_Symbol  Phenotype_Essentiality  Phenotype_Safety  \\\n",
       "0  Valaciclovir        GUK1                    16.0               3.0   \n",
       "1     Acyclovir        GUK1                    16.0               3.0   \n",
       "\n",
       "                     Pathways  \n",
       "0  Nucleotide interconversion  \n",
       "1  Nucleotide interconversion  "
      ]
     },
     "execution_count": 347,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sko_drugs.Pathways = sko_drugs.Pathways.apply(lambda x:x.split(';'))\n",
    "sko_drugs = sko_drugs.loc[:,['Drugs','Gene_Symbol','Pathways','Phenotype_Safety','Phenotype_Essentiality']]\n",
    "sko_drugs_ = explode(sko_drugs,['Pathways'],'')\n",
    "sko_drugs_ = sko_drugs_.loc[sko_drugs_.Pathways.apply(lambda x:len(x))>0,:]\n",
    "sko_drugs_ = sko_drugs_.drop_duplicates().reset_index(drop=True)\n",
    "sko_drugs_.to_csv('KO_data/SKO_db_Drugs_Unique_Pathways_tripartite.csv')\n",
    "sko_drugs_.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
